rfcs/45-stageless.md

```diff
@@ -0,0 +1,977 @@
+
+As a bonus, schedule construction will become completely order-independent as everything will be deferred until the end.
+
+```rust
+pub trait IntoConfiguredSystem<Params> {
+    fn configure(self) -> ConfiguredSystem;
+    fn configure_with(self, config: Config) -> ConfiguredSystem;
+    fn before<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystem;
+    fn after<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystem;
+    fn in_set(self, set: impl SystemLabel) -> ConfiguredSystem;
+    fn run_if<P>(self, condition: impl IntoRunCondition<P>) -> ConfiguredSystem;
+    fn named(self, name: impl SystemLabel) -> ConfiguredSystem;
+}
+
+pub trait IntoConfiguredSystemSet {
+    fn configure(self) -> ConfiguredSystemSet;
+    fn configure_with(self, config: Config) -> ConfiguredSystemSet;
+    fn before<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystemSet;
+    fn after<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystemSet;
+    fn in_set(self, set: impl SystemLabel) -> ConfiguredSystemSet;
+    fn run_if<P>(self, condition: impl IntoRunCondition<P>) -> ConfiguredSystemSet;
+}
+
+enum Order {
+    Before,
+    After
+}
+
+struct Config {
+    sets: HashSet<SystemLabelId>,
+    dependencies: Vec<(Order, SystemLabelId)>,
+    conditions: Vec<BoxedRunCondition>,
+}
+
+struct ConfiguredSystem {
+    system: BoxedSystem,
+    config: Config,
+    instance_name: SystemLabelId,
+}
+
+struct ConfiguredSystemSet {
+    set: SystemLabelId,
+    config: Config,
+}
+```
+
+### Schedule building errors
+
+When prototyping, you'll likely encounter schedule build errors.
+Most of these will point out a reason why some graph is unsolvable.
+
+There are two kinds of graphs the internal builder checks.
+
+The first kind consists of the `.in_set` relationships.
+These form a graph describing the hierachy of all your system sets.
+The second kind consists of the `.before` and `.after` relationships.
+Those form one or more dependency graphs (one for each set/schedule).
+Dependencies involving system sets will also be flattened into dependencies on the systems in those sets.
+
+So what are the errors?
+
+1. A dependency graph contains a loop or cycle.
+2. The hierarchical graph contains a loop or cycle.
+3. The hierarchical graph has a transitive edge. Example:
+    ```rust
+    app
+        .add_set(A);
+        .add_set(B.in_set(A));
+        // The next line causes a panic during graph solving.
+        // `A` cannot be a parent and grandparent to `C` at same time. 
+        .add_set(C.in_set(B).in_set(A))
+    ```
+4. You called `.in_set` with a label that belongs to a system, rather than a system set.
+5. (Optional) You have a dependency between two things that aren't siblings in a common set. That edge will not appear in either set's dependency graph, only in the flattened graph of an overarching set. This can lead to unwanted implicit ordering between systems in different sets.

```

Comment:
How is it possible to represent a situation like (5) when we multi-level graph representation, with a hierarchy graph, and a dependency graph for each system set? A cross-level dependency should start from a node in a dependency graph and end in a node of another dependency graph, which seems not possible to me, and would probably end up in a situation similar to (6), with unknown label (from the graph POV).

This, or maybe there is another representation *before* those two graphs that is able to represent this situation...

Reply:
A simple example would be, you're looking at two systems, A and B. A is under system set 1 and B is under system set 2. If you then say A depends on B or vice-versa, you'll hit (5).

This dependency *can* be valid, but said edge does not exist in the dependency graph of system set 1 or 2 (since each is only aware of one of the systems). The edge only appears in the flattened graph of any common ancestors of 1 and 2.

Reply:
Ok, thanks, I wrongly assumed that hierarchy and dependency graph generation was a destructive process, and those edges were lost forever, making those edges ultimately not appear in the flattened version. I need to read the `Systems` appendix section more carefully...